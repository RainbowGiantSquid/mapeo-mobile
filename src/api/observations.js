// @flow
import { jsonRequest } from './base';
import { keyBy } from 'lodash';
import { parseObservationResponse } from '../models/observations';
import querystring from 'querystring'

import type {
  Observation as ObservationType,
  ServerObservationResponse,
  ServerObservationCreate
} from '../types/observation';

class Observation {
  static list = (opts) =>
    jsonRequest({
      method: 'GET',
      route: '/observations?' + querystring.stringify(opts)
    }).map(observations => {
      var payload = observations.map(parseObservationResponse)
      return keyBy(payload, 'id')
    });

  static create = (observation: ServerObservationCreate) =>
    jsonRequest({
      method: 'POST',
      route: '/observations',
      body: observation
    }).map(parseObservationResponse);

  static update = (observation: ServerObservationResponse) =>
    jsonRequest({
      method: 'PUT',
      route: `/observations/${observation.id}`,
      body: observation
    }).map(parseObservationResponse);
}

export default Observation;
