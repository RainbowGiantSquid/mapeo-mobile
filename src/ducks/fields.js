// @flow

import update from 'immutability-helper';
import { keyBy } from 'lodash';
import { create } from '../lib/redux';

export const {
  type: FIELD_LIST,
  action: fieldList,
  reducer: fieldListReducer
} = create('FIELD_LIST', {
  success: (state, action) => {
    const newState = update(state, {
      fields: { $set: keyBy(action.payload, 'id') }
    });

    return newState;
  }
});

export default [fieldListReducer];
