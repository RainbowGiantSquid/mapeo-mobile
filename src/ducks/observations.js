// @flow
import { create } from '@src/lib/redux';
import update from 'immutability-helper';
import { keyBy } from 'lodash';
import { createSelector } from '@src/lib/selector';
import { StoreState } from '@types/redux';
import { Observation } from '@types/observation';

export const {
  type: OBSERVATION_LIST,
  action: observationList,
  reducer: observationListReducer,
} = create('OBSERVATION_LIST', {
  success: (state, action) => {
    const newState = update(state, {
      observations: { $set: keyBy(action.payload, 'id') },
    });

    return newState;
  },
});

export const selectContact = createSelector(
  [
    (state: StoreState, id: string): Observation => state.app.observations[id],
  ],
  (observation: Observation): Observation => observation,
  (observation: Observation, id: string): string => id,
);

export default [observationListReducer];
