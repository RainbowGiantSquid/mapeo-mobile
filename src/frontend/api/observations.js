import fixture from "./test_observations.json";

export default class ObservationsApi {
  constructor(baseUrl) {
    this.baseUrl = baseUrl;
  }
  list(cb) {
    setTimeout(() => cb(null, fixture), 2000);
  }
  update() {}
  create() {}
}
