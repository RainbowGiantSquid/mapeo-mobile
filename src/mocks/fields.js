// @flow

export const createField = (field?: any) => ({
  id: 'fieldId',
  name: 'fieldName',
  type: 'text',
  placeholder: 'this is some placeholder text',
  answered: false,
  answer: '',

  ...field
});
