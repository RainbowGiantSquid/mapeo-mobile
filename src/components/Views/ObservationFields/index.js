// @flow
import { connect } from 'react-redux';
import { NavigationActions } from 'react-navigation';
import type { Dispatch } from 'redux';
import { values } from 'lodash';
import { StoreState } from '../../../types/redux';
import { fieldList } from '../../../ducks/fields';
import { observationAdd, observationUpdate } from '../../../ducks/observations';
import ObservationFields from './ObservationFields';
import type { Props, StateProps, DispatchProps } from './ObservationFields';

function mapStateToProps(state: StoreState): StateProps {
  return {
    allFields: values(state.app.fields),
    selectedObservation: state.app.selectedObservation
  };
}

function mapDispatchToProps(dispatch: Dispatch): DispatchProps {
  return {
    addObservation: observation => dispatch(observationAdd(observation)),
    updateObservation: observation => dispatch(observationUpdate(observation)),
    goBack: () => dispatch(NavigationActions.back())
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(ObservationFields);
