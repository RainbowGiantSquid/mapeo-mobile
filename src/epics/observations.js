// @flow
import React from 'react';
import { Image } from 'react-native';
import type { ActionsObservable } from 'redux-observable';
import { OBSERVATION_LIST, observationList } from '../ducks/observations';
import { Action } from '../types/redux';
import { Observation } from '../types/observation';

export const observationListEpic = (
  action$: ActionsObservable<Action<string, Observation[]>>,
  store: any
) =>
  action$
    .ofType(OBSERVATION_LIST)
    .filter(
      action =>
        action.status === 'Start' &&
        store.getState().app &&
        store.getState().app.observations &&
        !store.getState().app.observations.length
    )
    .map(() =>
      observationList('', [
        {
          type: 'Old houses and farms',
          id: 1,
          lat: 0.106831547,
          lon: -77.42499473,
          link: 'link',
          created: new Date('2018-03-25T12:00:00Z'),
          name: 'Old houses and farms',
          notes: `Pedro and Maria's house`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/1-house.jpg')
            }
          ],
          icon: require('../images/categories/category_0.png'),
          mock: true
        },
        {
          type: 'Hunting sites',
          id: 2,
          lat: 0.088170932,
          lon: -77.44020885,
          link: 'link',
          created: new Date('2018-03-25T12:30:00Z'),
          name: 'Hunting sites',
          notes: `Don Pancho's hunting camp`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/4-camp.jpg')
            }
          ],
          icon: require('../images/categories/category_2.png'),
          mock: true
        },
        {
          type: 'Threat',
          id: 3,
          lat: 0.112411067,
          lon: -77.41378756,
          link: 'link',
          created: new Date('2018-02-03T17:30:00Z'),
          name: 'Threat',
          notes: `Oil spill 3 February 2018. Ruptured pipeline. Dead fish, oily water`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/5-threat.jpg')
            }
          ],
          icon: null,
          mock: true
        },
        {
          type: 'Medicinal plants',
          id: 4,
          lat: 0.09749025,
          lon: -77.43614404,
          link: 'link',
          created: new Date('2018-02-04T17:30:00Z'),
          name: 'Medicinal plants',
          notes: `Salola fruit plant - used in tea to heal inflamation`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/6-plant.jpg')
            }
          ],
          icon: require('../images/categories/category_6.png'),
          mock: true
        },
        {
          type: 'Waterfalls',
          id: 5,
          lat: 0.105818148,
          lon: -77.4408037,
          link: 'link',
          created: new Date('2018-03-04T17:30:00Z'),
          name: 'Waterfalls',
          notes: `Sacred waterfall, visited during vision quests for ancestors, for taking ayahuasca`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/7-waterfall.jpg')
            }
          ],
          icon: require('../images/categories/category_13.png'),
          mock: true
        },
        {
          type: 'Rivers and streams',
          id: 6,
          lat: 0.119004333,
          lon: -77.42374225,
          link: 'link',
          created: new Date('2018-03-24T17:30:00Z'),
          name: 'Rivers and streams',
          notes: `Point where pipeline crosses river`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/8-threat.jpg')
            }
          ],
          icon: require('../images/categories/category_0.png'),
          mock: true
        },
        {
          type: 'Historical sites',
          id: 7,
          lat: 0.082856811,
          lon: -77.4429836,
          link: 'link',
          created: new Date('2018-03-24T12:30:00Z'),
          name: 'Historical sites',
          notes: `There used to be a village here in the 1940s, it was abandoned after an epidemic of flu which killed most of the elders and children.`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/9-specialsite.jpg')
            }
          ],
          icon: require('../images/categories/category_17.png'),
          mock: true
        },
        {
          type: 'Gold',
          id: 8,
          lat: 0.108605553,
          lon: -77.41560494,
          link: 'link',
          created: new Date('2018-03-11T17:30:00Z'),
          name: 'Gold',
          notes: `Illegal gold miners 11 March 2018, excavators, dredges, open tailing pond, mercury.`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/10-threat.jpg')
            }
          ],
          icon: require('../images/categories/category_8.png'),
          mock: true
        }
      ])
    );

export default [observationListEpic];
