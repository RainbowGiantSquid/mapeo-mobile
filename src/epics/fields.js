// @flow
import React from 'react';
import type { ActionsObservable } from 'redux-observable';
import { FIELD_LIST, fieldList } from '../ducks/fields';
import { createField } from '../mocks/fields';
import type { Action } from '../types/redux';
import type { Field } from '../types/field';

const initialFields = [
  {
    name: 'Building type',
    type: 'text',
    placeholder: 'School/hospital/etc'
  },
  {
    name: 'Animal type',
    type: 'text',
    placeholder: 'What kind of animal'
  },
  {
    name: 'Name',
    type: 'text',
    placeholder: 'Name (if exists)'
  },
  {
    name: 'Items gathered',
    type: 'text',
    placeholder: 'What is collected here'
  },
  {
    name: 'Palm species',
    type: 'text',
    placeholder: 'What kind of palm'
  },
  {
    name: 'Owner',
    type: 'text',
    placeholder: 'Owner or user of the place'
  },
  {
    name: 'Plant species',
    type: 'text',
    placeholder: 'What kind of plant'
  },
  {
    name: 'Source',
    type: 'text',
    placeholder: 'Source of the data'
  },
  {
    name: 'Tree species',
    type: 'text',
    placeholder: 'What kind of tree'
  },
  {
    name: 'Type',
    type: 'text',
    placeholder: 'What kind of path'
  }
];

export const fieldListEpic = (
  action$: ActionsObservable<Action<string, Field[]>>
) =>
  action$
    .ofType(FIELD_LIST)
    .filter(action => action.status === 'Start')
    .map(() =>
      fieldList(
        '',
        initialFields.map((f, i) =>
          createField({
            name: f.name,
            id: i.toString(),
            type: f.type,
            placeholder: f.placeholder
          })
        )
      )
    );

export default [fieldListEpic];
