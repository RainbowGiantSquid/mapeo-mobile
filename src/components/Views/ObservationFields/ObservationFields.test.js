// @flow
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import ObservationFields from './ObservationFields';
import { createObservation } from '../../../mocks/observations';
import { createField } from '../../../mocks/fields';

jest.mock('../../Base/FieldInput', () => () => null);

describe('ObservationFields tests', () => {
  const observation = createObservation();
  const isFocused = () => true;
  const addListener = () => true;
  const allFields = [createField()];

  test('snapshot', () => {
    const tree = renderer.create(
      <ObservationFields
        allFields={allFields}
        navigation={{ isFocused, addListener }}
        selectedObservation={observation}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
