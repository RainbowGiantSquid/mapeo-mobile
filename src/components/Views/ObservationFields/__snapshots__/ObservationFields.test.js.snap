// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObservationFields tests snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View />
  </RCTScrollView>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, .8)",
        "flexDirection": "row",
        "height": 50,
        "paddingHorizontal": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#ffffff",
          "fontWeight": "400",
          "paddingRight": 25,
          "textAlign": "center",
        }
      }
    >
      0 of 1 answered
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingTop": 2,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#ffffff",
            "borderRightColor": "rgba(0, 0, 0, .85)",
            "borderWidth": 0.5,
            "height": 8,
            "width": 575,
          }
        }
      />
    </View>
  </View>
</View>
`;
